<h2><center>Assignment 2</center></h2>
<h3><center>Programming for Data Science 2026</center></h3>
<b><center>Deadline: 14:00, March 3, 2026</center></b>

- The exercise will be marked as **passed** if you get **at least 10/16** points.

- Exercises must be handed in via **ILIAS** (Homework assignments). Submit your work as a **compressed (.zip)** file containing **one** `.py` **or** `.ipynb` file with **all exercises**.

- The name of **both** the `.zip` file and the `.py`/`.ipynb` file must be the *SurnameName* of the **two group members**, separated by an underscore.  
  Example: Tim Tabellen and Lara Lambda → `TabellenTim_LambdaLara.zip`  
  (The `.py`/`.ipynb` file must have the same name.)

- Use **comments** to explain your code and demonstrate that you understand the solutions and can discuss them.

- You are not expected to collaborate outside your group. Submitting other groups’ code as your own will result in **0 points**.

- For general questions about the lecture content, assignments, or the exam, please use the **ILIAS forum**.

- For individual questions about lecture content and the exam, contact: *roland.widmer@students.unibe.ch*  

- For individual questions about the exercises or grading, contact: *julien.brunner@students.unibe.ch* or *aline.steiner@students.unibe.ch*

#### Exercise 1: Lists *(3 points)*

In [1]:
from statistics import mean

from numpy.ma.core import sqrt

list_1 = [1, 2, 3]
list_2 = list_1

list_2.append(4)

print("List 1:", list_1)
print("List 2:", list_2)

List 1: [1, 2, 3, 4]
List 2: [1, 2, 3, 4]


Explain why both lists contain 4, even though we do not have code directly involving `list_1`.

**Your answer**: Appending is a mutation operation. list_1 and list_2 are references to the same object. When you mutate the object, all references will show the updated object. In this case, with the appended 4.

Demonstrate your explanation by using `id(...)` and comments.

In [2]:
print(f'Are ID of list_1 and list_2 equals : {id(list_1) == id(list_2)}')
# id retrieves the integer value of the memory address. When we compare both list_1 and list_2's id, we find that they are the same value hence showing that they are references to the same object

Are ID of list_1 and list_2 equals : True


Explain whether this phenomenon also occurs with immutable types.

**Your answer**: This does not happen with immutable types

Say you do \
a = 10 \
b = a \
a = 5

This is because when you create a pointer to the object, say \"a\" from above, and then you reassign it, (as it can only be reassigned and not changed: it is immutable), it will not affect the value of another pointer, because b is still pointing to the original object. So here, b == 10.

#### Exercise 2: Filtering *(2 points)*

Implement a function `odd_numbers` that accepts a list of integers and returns a list containing only the odd numbers of this list.

Example: `print(odd_numbers([1, 2, 3]))` should output `[1, 3]`

In [3]:
def odd_numbers(list):
    out = []
    for number in list:
        # if number % 2 == 1:   # Rohit version
        if number & 1 == 1:  # Mario version : bitwise comparison is faster than modulo
            out.append(number)
    return out


print(odd_numbers([1, 2, 3, 4, 5]))

[1, 3, 5]


#### Exercise 3: Conditions *(2 points)*

Implement a function `coffee_status` with two integers `cups_today` and `hours_slept` as input. Based on the following rules, it should print one statement:

- If `hours_slept` < 7 print "Tired" but only if `cups_today` is 0
- But if `cups_today` is > 5 print "Too much caffeine" and not "Tired"
- Otherwise, print "OK".

Hint: Test your function with different values. Does it behave as expected?

In [4]:
def coffee_status(cups_today, hours_slept):
    if hours_slept < 7 and cups_today == 0:
        print("Tired")
    elif cups_today > 5:
        print("Too much caffeine")
    else:
        print("OK")


coffee_status(7, 3)
coffee_status(2, 5)
coffee_status(0, 6)
# behaves like expected

Too much caffeine
OK
Tired


#### Exercise 4: Lottery *(5 points)*

- In this exercise, we use the `randint` function, which can be used to generate pseudorandom integers.
- Implement a function `gamble` that accepts an inputs `fee`.
- The function should initialize a variable `spending` with 0.
- Then, the `fee` should be deducted from `spending` and 3 random integers between 0 and 9 should be generated.
- If all three numbers are `6`, the program stops (victory).
- Otherwise, it should continue by deducting the fee again from `spending` and by generating another set of 3 random integers.
- Use `print` in every iteration to ouput the current state (the 3 random numbers and the current `spending`). In the end, the program should return the (negative) `spending`.

In [5]:
import random


def gamble(fee, verbose=False):
    spending = 0

    while True:
        spending -= fee
        nums = [random.randint(1, 9) for _ in range(3)]

        if verbose:
            print(f"Numbers: {nums} , Spent: {spending}")

        if nums == [6, 6, 6]:
            break
    return spending


gamble(2, True)

Numbers: [6, 9, 6] , Spent: -2
Numbers: [3, 3, 3] , Spent: -4
Numbers: [9, 5, 1] , Spent: -6
Numbers: [1, 9, 1] , Spent: -8
Numbers: [4, 3, 9] , Spent: -10
Numbers: [6, 6, 7] , Spent: -12
Numbers: [7, 4, 3] , Spent: -14
Numbers: [3, 2, 6] , Spent: -16
Numbers: [1, 6, 7] , Spent: -18
Numbers: [8, 8, 9] , Spent: -20
Numbers: [7, 9, 5] , Spent: -22
Numbers: [3, 1, 3] , Spent: -24
Numbers: [3, 1, 3] , Spent: -26
Numbers: [7, 8, 4] , Spent: -28
Numbers: [9, 3, 2] , Spent: -30
Numbers: [1, 9, 2] , Spent: -32
Numbers: [9, 3, 8] , Spent: -34
Numbers: [9, 7, 9] , Spent: -36
Numbers: [8, 7, 6] , Spent: -38
Numbers: [3, 1, 5] , Spent: -40
Numbers: [7, 9, 1] , Spent: -42
Numbers: [3, 5, 2] , Spent: -44
Numbers: [7, 8, 2] , Spent: -46
Numbers: [7, 7, 6] , Spent: -48
Numbers: [3, 5, 2] , Spent: -50
Numbers: [1, 9, 8] , Spent: -52
Numbers: [8, 9, 2] , Spent: -54
Numbers: [9, 6, 7] , Spent: -56
Numbers: [4, 3, 7] , Spent: -58
Numbers: [9, 8, 5] , Spent: -60
Numbers: [3, 1, 3] , Spent: -62
Numbers: [5,

-670

- Create a loop and call the function `gamble` many times (e.g., 1000) with a fee of 5 CHF.
- Change the `gamble` function to make the many calls of `print` optional by adding an argument `verbose`. If `gamble(..., verbose=False`), there should be no prints.
- Add the return value of the `gamble` function to a list.
- Calculate the average spending.

As a casino owner, what is the maximum jackpot prize that you can award to winners for a win without making loss in the long run?

In [6]:
import numpy as np

n = 1000
fee = 5

spendings = []
for i in range(n):
    spendings.append(gamble(fee))

mean = sum(spendings) / n
print(f'Average spending of {mean} on {n} runs with fee = {fee}')

Average spending of -3742.03 on 1000 runs with fee = 5


Given an average spending of -3742.03 and a fixed fee of 5, setting the maximum jackpot equal to the average spending ensures a sustainable model in the long run. While short-term deviations in spending may result in temporary profits or losses, these fluctuations will eventually average out to a zero balance over time.

#### Exercise 5: Number divisors *(4 points)*

A divisor of a number, is another number that divides the first number without reminder.

Example: 4 is a divisor of 20. 1 and the number itself are always divisors. 

Write a function `find_divisors` that takes a number in input and returns *all* its divisors.

Example:
```python 
find_divisors(20)  # returns [1, 2, 4, 5, 10, 20]
find_divisors(7)  # returns [1, 7]
```

In [7]:
from math import sqrt


def find_divisors(number):
    divisors = set()  # Using set so it sort it nicely
    # Ranging from 1 to sqrt bc we know that bigger divisors would result in a number bigger than the number, other half is covered by the corresponding divisor
    for i in range(1, int(sqrt(number)) + 1):
        if number % i == 0:
            divisors.add(i)
            corresponding_divisor = number // i
            if i != corresponding_divisor:  # Add the corresponding divisors if not the same
                divisors.add(corresponding_divisor)
    return divisors


print(find_divisors(20))

{1, 2, 4, 5, 10, 20}


Implement a function `find_common_divisors` that accepts to integers and returns a sorted list of common divisors (high divisors first).

Hints: Use the function `find_divisors` from above. Sets can be helpful to implement list intersections.

In [8]:
def find_common_divisors(a, b):
    return set(find_divisors(a)) & set(find_divisors(b))

find_common_divisors(56, 48)

{1, 2, 4, 8}